File: 0768-partition-labels/0768-partition-labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        int n = s.length();
        vector<int> ans;
        array<int, 26> end;

        for (int i = 0; i < n; ++i) {
            int key = s[i] - 'a';
            end[key] = i;
        }
        int prev = -1, last = -1;
        for (int i = 0; i < n; ++i) {
            int key = s[i] - 'a';
            if (end[key] > last)
                last = end[key];
            if (i == last) {
                ans.push_back(i - prev);
                prev = i;
            }
        }
        return ans;
    }
};

File: 0768-partition-labels/0768-partition-labels.js
/**
 * @param {string} s
 * @return {number[]}
 */
var partitionLabels = function(s) {
    let n = s.length;
    const ans = []
    const arr = new Array(26).fill(0);

    for (let i = 0; i < n; ++i) {
        let idx = s.charCodeAt(i) - 97;
        arr[idx] = i;
    }
    let prev = -1, last = -1;
    for (let i = 0; i < n; ++i) {
        let idx = s.charCodeAt(i) - 97;
        if (arr[idx] > last)
            last = arr[idx];
        if (i == last) {
            ans.push(i - prev);
            prev = i;
        }
    }
    return ans;
};